CODING_GUIDELINES.md

```diff
@@ -1,13 +1,140 @@
 # Coding Guidelines
-This document describes coding guidelines.
+This document describes our coding guidelines.
+One of the main points of this project is, to keep code readable.
+This is why we enforce the following rules.
+If you are a contributor to this project, please follow these rules.
+If you disagree with one or more of them, please open an issue and provide solid description why we should change them.
+Many of these rules are enforced by static code analysis, which you can run locally with `make lint`.
 
-## panic
-No `panic` must be used.
-It endangers the stability of the whole application.
+## Files
+When creating a new package `mypackage`, we aim for the following files in the folder.
+* `doc.go` only consists of a package-documentation comment and a package declaration.
+* `example_test.go` contains example functions that show how to use your package.
+  Use of `package mypackage_test` is mandatory here. Be aware that example functions are testable in Go.
+* `mypackage.go` contains the entry point of this package
+  For ease of readability, no other file should contain an entry point to your package (except structs).
+* `mypackage_test.go` contains tests for your package.
+  If you like, you can use the `package mypackage_test` in that file, however, this is up to you.
+* `error.go` contains sentinel errors.
+  For possible content, see section _Sentinel errors_ below.
+* Any other file in your package that is necessary.
+  If you use a lot of structs with methods, try to place each struct in a separate file, that is named like the struct, but in lower-snake-case (`hello_world`).
+
+## Code
+We want to keep our code readable (that's one of the main points of this project) and clean.
+Many of the things we expect from code, is ensured by our internal analyzers (`internal/tool/analysis`).
+However, not everything may be covered by those.
+This is why we rely on you to follow the below rules for our code.
+
+### Ordering
+We want to keep the following structure in any non-test `.go`-file.
+
+1. package declaration (no package comment, use a file `doc.go` for that)
+2. import declarations
+3. const declarations
+4. var declarations
+5. interface declarations
+6. struct declarations (followed by constructor functions)
+7. exported methods
+8. unexported methods
+9. exported functions
+10. unexported functions
+
+If there are multiple independent structs in one file, you should probably split the file into two or more.
+
+### Constructor functions
+Use the `&struct` syntax for initialization whenever possible.
```

Comment:
why pointer initialisation made mandatory ??

Reply:
Constructors always need to return the pointer instead of the variable.

Reply:
If we are not using constructor, like using variable in a fuction there we can use as variable ?
Why constructor need to always return the pointer reasons for that ?

Reply:
Constructors are usually named `NewXXX`, and should return a pointer, just as `new(T)` returns a pointer. `make(T)` does not.

Reply:
When you create a variable somewhere else, you can use a different method, if it makes sense. This paragraph only regards constructor functions.